src/rgw/rgw_sync.cc

```diff
@@ -2226,3 +2226,124 @@ int RGWCloneMetaLogCoroutine::state_store_mdlog_entries_complete()
+{
+  connection_map connections;
+  for (auto& g : zonegroups) {
+    for (auto& z : g.second.zones) {
+      std::unique_ptr<RGWRESTConn> conn{
+        new RGWRESTConn(store->ctx(), store, z.first, z.second.endpoints)};
+      connections.emplace(z.first, std::move(conn));
+    }
+  }
+  return connections;
+}
+
+struct TrimEnv {
+  RGWRados *const store;
+  RGWHTTPManager *const http;
+  int num_shards;
+  const std::string& zone;
+  Cursor current; //< cursor to current period
+
+  TrimEnv(RGWRados *store, RGWHTTPManager *http, int num_shards)
+    : store(store), http(http), num_shards(num_shards),
+      zone(store->get_zone_params().get_id()),
+      current(store->period_history->get_current())
+  {}
+};
+
+struct MasterTrimEnv : public TrimEnv {
+  connection_map connections; //< peer connections
+  std::vector<rgw_meta_sync_status> peer_status; //< sync status for each peer
+  /// last trim marker for each shard, only applies to current period's mdlog
+  std::vector<std::string> last_trim_markers;
+
+  MasterTrimEnv(RGWRados *store, RGWHTTPManager *http, int num_shards)
+    : TrimEnv(store, http, num_shards),
+      last_trim_markers(num_shards)
+  {
+    auto& period = current.get_period();
+    connections = make_peer_connections(store, period.get_map().zonegroups);
+    connections.erase(zone);
+    peer_status.resize(connections.size());
+  }
+};
+
+} // anonymous namespace
+
+
+/// the meta master zone decides which entries are safe to trim by querying
+/// the sync status of all peer zones. an mdlog entry is only safe to trim if
+/// all peers indicate a sync status marker greater than or equal to that
+/// entry. similarly, entire mdlog periods can be safely removed once all peer
+/// zones report a realm_epoch higher than that mdlog period. this logic
+/// guarantees that no mdlog entries will be trimmed while other zones may try
+/// to consume them
+class MetaMasterTrimCR : public RGWCoroutine {
+  MasterTrimEnv& env;
+  int ret{0};
+
+ public:
+  MetaMasterTrimCR(MasterTrimEnv& env)
+    : RGWCoroutine(env.store->ctx()), env(env)
+  {}
+
+  int operate() override;
+};
+
+int MetaMasterTrimCR::operate()
+{
+  reenter(this) {
+    if (env.connections.empty()) {
+      ldout(cct, 4) << "no peers, exiting" << dendl;
+      return set_cr_done();
+    }
+
+    ldout(cct, 10) << "fetching sync status for zone " << env.zone << dendl;
+    yield {
+      // query mdlog sync status from peers
+      rgw_http_param_pair params[] = {
+        { "type", "metadata" },
+        { "status", nullptr },
+        { nullptr, nullptr }
+      };
+
+      auto p = env.peer_status.begin();
+      for (auto& c : env.connections) {
+        ldout(cct, 20) << "query sync status from " << c.first << dendl;
+        using StatusCR = RGWReadRESTResourceCR<rgw_meta_sync_status>;
+        auto conn = c.second.get();
+        spawn(new StatusCR(cct, conn, env.http, "/admin/log/", params, &*p),
+              false);
```

Comment:
@cbodley some other place we have a window of operations, so that we don't spawn more than X concurrent crs at once. Maybe here is a good candidate for doing the same?